src/pydicom/pixels/decoders/base.py

```diff
@@ -0,0 +1,1980 @@
+            return value
+
+        raise AttributeError("No value for 'columns' has been set")
+
+    def decode(self, index: int) -> bytes | bytearray:
+        """Decode the frame of pixel data at `index`.
+
+        Parameters
+        ----------
+        index : int
+            The index of the frame to be decoded.
+
+        Returns
+        -------
+        bytes | bytearray
+            The decoded frame of pixel data.
+        """
+        # For encapsulated data `self.src` should not be memoryview to avoid
+        #   creating a duplicate object in memory by the encapsulation functions
+        src = get_frame(
+            self.src,
+            index,
+            number_of_frames=self.number_of_frames,
+            extended_offsets=self.extended_offsets,
+        )
+
+        return self._decode_frame(src)
+
+    def _decode_frame(self, src: bytes) -> bytes | bytearray:
+        """Return a decoded frame of pixel data.
+
+        Parameters
+        ----------
+        src : bytes
+            An encoded frame of pixel data to be passed to the decoding plugins.
+
+        Returns
+        -------
+        bytes | bytearray
+            The decoded frame.
+        """
+        # If self._previous is not set then this is the first frame being decoded
+        # If self._previous is set and then the previously successful decoder
+        #   has failed while decoding a frame and we are trying the other decoders
+        failure_messages = []
+        for name, func in self._decoders.items():
+            try:
+                # Attempt to decode the frame
+                frame = func(src, self.options)
+
+                # Decode success, if we were previously successful then
+                #   warn about the change to the new decoder
+                if hasattr(self, "_previous") and self._previous[1] != func:
+                    warn_and_log(
+                        f"The decoding plugin has changed from '{self._previous[0]}' "
+                        f"to '{name}' during the decoding process - you may get "
+                        f"inconsistent inter-frame results, consider passing "
+                        f"'decoding_plugin=\"{name}\"' instead"
+                    )
+
+                self._previous = (name, func)
+                return frame
+            except Exception as exc:
+                LOGGER.exception(exc)
+                failure_messages.append(f"{name}: {exc}")
+
+        messages = "\n  ".join(failure_messages)
+        raise RuntimeError(
+            "Unable to decode as exceptions were raised by all available "
+            f"plugins:\n  {messages}"
+        )
+
+    def del_option(self, name: str) -> None:
+        """Delete option `name` from the runner."""
+        if name in ("transfer_syntax_uid", "pixel_keyword"):
+            raise ValueError(f"Deleting '{name}' is not allowed")
+
+        if name in self._opts:
+            del self._opts[name]  # type: ignore
+
```

Comment:
Very minor suggestion: could one-line this with `self._opts.pop(name, None)` which silently ignores if missing.

Reply:
Done